### PR TITLE
PP-6184: Simplify smartpay status mapping

### PR DESCRIPTION
**Pull request by oswaldquek**
The old StatusMapper was only used by SmartpayStatusMapper, which was using
generics in an unnecessary way which made it hard to understand what was going
on. Usage of the `Pair` class in various places compounded this problem. Usages
of `Pair` is replaced by a concrete SmartpayStatus.

## WHAT YOU DID
_A brief description of the pull request:_

## How to test

- How should it be reviewed? 
- What feedback do you need? 
- If manual testing is needed, give suggested testing steps.

### Logging

- [ ] only emit log lines at ERROR level which require immediate attention from a support engineer. These will trigger a zendesk alert.

### Documentation

- [ ] Updated README.md for any of the following ?

* Introduced any new environment variables / removed existing environment variable
* Added new API / updated existing API definition

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayStatusMapper.java [hunk 1]
```diff
@@ -11,33 +21,80 @@
...
+
+        if (status instanceof ChargeStatus) {
+            return new MappedChargeStatus((ChargeStatus) status);
+        }
+
+        if (status instanceof RefundStatus) {
+            return new MappedRefundStatus((RefundStatus) status);
+        }
+
+        return new UnknownStatus();
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/gateway/smartpay/SmartpayStatusMapper.java]: I think everything from `Optional<Status> statusMaybe = …` could be replaced by:

```java
return statusMap
        .flatMap(StatusMap::getToStatus)
        .map(status -> {
            if (status instanceof ChargeStatus) {
                return new MappedChargeStatus((ChargeStatus) status);
            }
            if (status instanceof RefundStatus) {
                return new MappedRefundStatus((RefundStatus) status);
            }
            return new UnknownStatus();
        })
        .orElseGet(IgnoredStatus::new);
```

File: src/test/java/uk/gov/pay/connector/gateway/smartpay/SmartpayStatusMapperTest.java [hunk 2]
```diff
@@ -16,70 +13,70 @@
...
+        assertThat(status.getType()).isEqualTo(InterpretedStatus.Type.UNKNOWN);
     }
 
     @Test
     public void shouldReturnEmptyWhenStatusIsIgnored() {
-        InterpretedStatus status = SmartpayStatusMapper.from(Pair.of("AUTHORISATION", true));
+        InterpretedStatus status = SmartpayStatusMapper.from(new SmartpayStatus("AUTHORISATION", true));
 
-        assertThat(status.getType(), is(InterpretedStatus.Type.IGNORED));
+        assertThat(status.getType()).isEqualTo(InterpretedStatus.Type.IGNORED);
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/gateway/smartpay/SmartpayStatusMapperTest.java:79]: I like the way that there’s no logic changes to these tests so we can ensure the behaviour hasn’t changed.

**Review by alexbishop1 — DISMISSED** [review]
Nice simplification. One suggestion, which you can take or leave.

**Comment by oswaldquek** [comment]
@alexbishop1 I took your suggestion, could you approve again please?

**Review by alexbishop1 — APPROVED** [review]

**Pull request merged**